Support metric dependent formulas

cwasicki opened the pull request:
This supports having defining different formulas for each metric in the config file. An example for different formulas would be power values, which should be summed, and SoC, where a sum doesn't make sense. Instead a capacity-weighted mean can be used for this.

Moreover, the metric name can be passed to the `cids` method to support logic depending on the metric name. Since the method is used to allow getting all CIDs that are required to calculate certain component type metrics, the CIDs can differ depending on the metric that is calculated.

Without passing a metric, the logic of the method is unchanged. If a metric was passed, the method extracts the CIDs from the formula. The formula can be either a user-defined formula or a default formula if available.

cwasicki committed:
Update to more flexible formula config

This supports having defining different formulas for each metric in the
config file. An example for different formulas would be power values,
which should be summed, and SoC, where a sum doesn't make sense. Instead
a capacity-weighted mean can be used for this.

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki committed:
Support getting metric-specific CIDs

The metric name can be passed to the `cids` method to support logic
depending on the metric name. Since the method is used to allow getting
all CIDs that are required to calculate certain component type metrics,
the CIDs can differ depending on the metric that is calculated.

Without passing a metric, the logic of the method is unchanged. If a
metric was passed, the method extracts the CIDs from the formula. The
formula can be either a user-defined formula or a default formula if
available.

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki committed:
Update release notes

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

flora-hofmann-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -2,12 +2,11 @@
 
-- If your code references `"load"` as a component type, update it to `"consumption"`.
-- Made the `MicrogridConfig` reader tolerant to missing `ctype` fields, allowing collection of incomplete microgrid configs.
+* Formula configs are now defined per metric to support different formulas for each metric in the same config file. A
```
There's a rogue 'A' at the end of the sentence.

flora-hofmann-frequenz commented on src/frequenz/lib/notebooks/config.py:
```diff
@@ -30,20 +31,52 @@ class ComponentTypeConfig:
+                raise ValueError(
+                    f"{metric} does not have a formula for {self.component_type}"
+                )
+            pattern = r"#(\d+)"
```
maybe a little comment would help here to explain the pattern for an example #123 or something. Not sure if everyone is regex literate.

flora-hofmann-frequenz reviewed (dismissed):
Just two smaller comments that could be worked in, otherwise LGTM.

cyiallou reviewed (commented):
Besides the two comments from Flora, it looks good to me.

cwasicki committed:
Add default formula for SoC

This uses a simple average of all SoC values which ignores the impact
of different battery capacities.

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki commented:
@flora-hofmann-frequenz @cyiallou Updated according to your comments and fixed some issues to pass the new CI.

cyiallou commented on RELEASE_NOTES.md:
```diff
@@ -2,12 +2,14 @@
 ## Upgrading
 
-- If your code references `"load"` as a component type, update it to `"consumption"`.
-- Made the `MicrogridConfig` reader tolerant to missing `ctype` fields, allowing collection of incomplete microgrid configs.
```
The bullet point "Made the `MicrogridConfig`" should not be deleted. There was a mistake in a previous PR (see commit: https://github.com/frequenz-floss/frequenz-lib-notebooks/commit/f6939c6e1d763dea6cfd9d347b3ebd8fb0e73019) where the first bullet point, which belongs to the previous release, got deleted and added back with the referenced commit.

cyiallou commented on src/frequenz/lib/notebooks/config.py:
```diff
@@ -240,11 +235,13 @@ def formula(self, component_type: str, metric: str) -> str:
-
-        if not cfg.has_formula_for(metric):
+        if cfg.formula is None:
+            raise ValueError(f"No formula set for {component_type}")
```
Could we add all the different `ValueError`s to the docstring for completeness please? :)

cyiallou reviewed (changes requested):
The logic looks good to me. Just a minor comment on the release notes and some cosmetic changes.

cwasicki commented:
@cyiallou Fixed
